### PR TITLE
Implement intelligent cache/origin selection algorithm

### LINKED ISSUE #1461: Intelligent Director cache selection logic
See details in https://github.com/PelicanPlatform/pelican/discussions/1198

### PR DESCRIPTION
**Pull request by haoming29**
Closes #1461 

This is pending on #1454 to get required object availability data for the sorting algorithm

### CODE REVIEW

**Comment by haoming29** [comment]
@bbockelm one question came up when I'm implementing this. Should we do the same algorithm to `redirectToOrigins`? Right now they share the same sorting function, but we can separate them out if this is not desired.

File: director/sort.go
Comment (jhiemstrawisc) [inline director/sort.go:220]: There's a typo in the function name -- it should be `gatedHalvingMultiplier` (switched v/l)

File: director/sort_algorithms_test.go
Comment (jhiemstrawisc) [inline director/sort_algorithms_test.go:66]: Is this safe if someone accidentally passes a negative value? Can we make this a `uint`?

File: director/sort_algorithms.go
Comment (jhiemstrawisc) [inline director/sort_algorithms.go:109]: Why return here? If we tell the sort algorithm "don't give me more than 10 back" but there are only 5 candidates, shouldn't we still sort those 5 and return them?

File: docs/parameters.yaml
Comment (jhiemstrawisc) [inline docs/parameters.yaml:1271]: How about something like "Sorts caches according to stochastically-generated weights that consider a combination of factors, including a cache's distance from the client, its IO load, and whether the cache already has the requested object."

On a side note, I don't love the term "smart" because I think this is the kind of thing Miron would have issues with. What about something like "balanced", "adaptive" or "dynamic"? Feel free to disagree to disagree with me on this, I'll defer to you.

**Review by jhiemstrawisc — CHANGES_REQUESTED** [review]
Excellent, this looks like it was a fun one! A few minor comments inline, and a few bigger issues listed here.

When I tested locally, I served an OSDF director with `Director.CacheSortMethod: "smart"`. Curling the director with `curl -v -k https://<hostname>/chtc/PUBLIC/foo` resulted in a broken list of responses where many of the server hostnames have disappeared:
```
< link: <https:///chtc/PUBLIC/foo>; rel="duplicate"; pri=1; depth=1, <https:///chtc/PUBLIC/foo>; rel="duplicate"; pri=2; depth=1, <https:///chtc/PUBLIC/foo>; rel="duplicate"; pri=3; depth=1, <http://osg-kansas-city-stashcache.nrp.internet2.edu:8000/chtc/PUBLIC/foo>; rel="duplicate"; pri=4; depth=1, <http://stash.farm.particle.cz:8000/chtc/PUBLIC/foo>; rel="duplicate"; pri=5; depth=1, <http://osdf-cache.sprace.org.br:8000/chtc/PUBLIC/foo>; rel="duplicate"; pri=6; depth=1
< location: https:///chtc/PUBLIC/foo
< x-pelican-namespace: namespace=/chtc/PUBLIC, require-token=false
< x-pelican-token-generation: max-scope-depth=0
<a href="https:///chtc/PUBLIC/foo">Temporary Redirect</a>.
```
In all tests, it's the first three servers that are missing their hostnames. This is probably coming from the configured `candidateLimit` and some mishandling of the returned candidates, but I haven't gone that far to investigate.

Another issue I encountered is that when I set the sorting method to "distanceAndLoad" and performed the same curl, the Director gets a private IP address and correctly reports "GeoIP Resolution of the address 172.17.0.4 resulted in the null lat/long." When this happens, I get what appears to be a deterministic response with the same ordering:
```
< link: <http://amst-osdf-xcache01.es.net:8000/chtc/PUBLIC/foo>; rel="duplicate"; pri=1; depth=1, <http://dtn-pas.bois.nrp.internet2.edu:8000/chtc/PUBLIC/foo>; rel="duplicate"; pri=2; depth=1, <http://stash-cache.osg.chtc.io:8000/chtc/PUBLIC/foo>; rel="duplicate"; pri=3; depth=1, <http://dtn-pas.cinc.nrp.internet2.edu:8000/chtc/PUBLIC/foo>; rel="duplicate"; pri=4; depth=1, <http://osg-stash-sfu-computecanada-ca.nationalresearchplatform.org:8000/chtc/PUBLIC/foo>; rel="duplicate"; pri=5; depth=1, <http://dtn-pas.denv.nrp.internet2.edu:8443/chtc/PUBLIC/foo>; rel="duplicate"; pri=6; depth=1
< location: http://amst-osdf-xcache01.es.net:8000/chtc/PUBLIC/foo
< x-pelican-namespace: namespace=/chtc/PUBLIC, require-token=false
< x-pelican-token-generation: max-scope-depth=0
<a href="http://amst-osdf-xcache01.es.net:8000/chtc/PUBLIC/foo">Temporary Redirect</a>.
```
Shouldn't there still be something stochastic here as observed when using `distance`, even if we know the load values?

**Comment by haoming29** [comment]
@jhiemstrawisc Regarding your top-level comments:
1. Yes! Thanks for catching that. There's a bug in the sorting code where I preemptively created a slice with size of `cadidateLimit` then did the `append`, which basically means the first three items are empty and not useful. This is fixed
2. This behavior is expected based on the current implementation, which is that if the client IP resolves to nothing, we basically ignores any "distance' related weights. In practice, this means that:
* For `distance`, we do random weight
* For `distanceAndLoad`, we ignore the distance part and only takes load into consideration and generate the weight based on load only.
* For `smart`, similar to  `distanceAndLoad`, but we still consider `availability` as another weight.

The reason you always receive the same ordering is due to a known bug https://github.com/PelicanPlatform/pelican/issues/1538 where the server IO load are not available among cache servers (bummer!)

File: director/sort.go
Comment (haoming29) [reply director/sort.go:71]: This is a good question for @bbockelm

File: director/sort_algorithms_test.go
Comment (haoming29) [reply director/sort_algorithms_test.go:66]: Alright, I will take `max(len(sm), maxOut)` as the final `maxOut` value to be safe.

File: director/sort_algorithms.go
Comment (haoming29) [reply director/sort_algorithms.go:109]: See reply above

File: docs/parameters.yaml
Comment (haoming29) [reply docs/parameters.yaml]: Naming is always hard! I like "adaptive" and really open to any suggestions

**Review by turetske — COMMENTED** [review]
Solely looking at the changes since your last review, it looks good. I'll approve once we figure out the issue with the build so that the tests start passing.

**Review by turetske — APPROVED** [review]
LGTM

**Review by jhiemstrawisc — APPROVED** [review]

**Pull request merged**
